tests/functional/test_stub.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+from tests import unittest
+import mock
+
+import botocore
+from botocore.stub import Stubber
+from botocore.credentials import Credentials
+from botocore.exceptions import StubResponseError, ClientError
+from botocore.loaders import Loader
+from botocore import hooks
+import botocore.client
+import botocore.retryhandler
+import botocore.translate
+
+
+class TestStubber(unittest.TestCase):
+    def setUp(self):
+        credentials = Credentials('access-key', 'secret-key')
+
+        with mock.patch('botocore.client.EndpointCreator') \
+                as endpoint_creator_patch:
+            endpoint_creator = endpoint_creator_patch.return_value
+            self.endpoint = mock.Mock()
```

Comment:
So for this endpoint mocking, let's just move it to the unit test. These should not be in the functional tests. What I would do for these functional test is remove all instances of asserting the endpoint was hit and compact them into a few unit tests that do all of patching and then ensure that the endpoint is not hit. In the end, I would prefer no patches or mocks being used.

Everything else in the functional test look fine though.